Snakefile

```diff
@@ -18,14 +18,10 @@ from pathlib import Path
 
 HTTP = HTTPRemoteProvider()
 
-if "config" not in globals() or not config:  # skip when used as sub-workflow
-    if not exists("config.yaml"):
-        copyfile("config.tutorial.yaml", "config.yaml")
-
-    configfile: "config.yaml"
-
```

Comment:
Are we sure about this?
I'm wondering that it may be good to preserve.
What do you think?
I like though specifying the three configfile: as proposed but probably within the if?

Reply:
Hey @davide-f, thanks for your reply! we are getting there :) This is indeed one change that I am not sure about. The reason why I put this in is that I think the workflow should also be tested against the `config.default.yaml` settings. Before, the CI was "only" testing the tutorial config as a basis. This leads to the artefact that one always has to secure manually that the default config is in line with the tutorial config. The proposed approach is the same as we have it in pypsa-eur and uses 

- config.default.yaml as level-zero defaults
- add configs/bundle_config.yaml which extends the default (no overrides)
- use incremental config changes from the user via `config.yaml`. Here the user can only set the keys which should be different from the `config.default.yaml`. In a python function, that would translate to setting a keyword-argument which already has a default to something else. So the `config.yaml` could be empty, and the workflow would still run.   

The structure ensure that the user always has a stable set of default configs. Whenever there is an update to the default, the user does not need to copy it to `config.yaml` and insert their custom settings manually, it happens automatically.  

However, it leads to one or two more needed fixes in the workflow. For example, the CI is breaking now as the solver_options are not fully compatible. Snakemake is currently merging the solver options from the default solver with the ones from the tutorial solver. 

Two options:

1. Go full in -> make an adoption of the solver_options as we have it in pypsa-eur (already happened and will be happening anyway through #796 )
2. Revert this setting and introduce it at a later point (which I would strongly encourage to do)

what to you think?

Reply:
Hello @FabianHofmann! Sounds like a well-elaborated and quite an elegant solution.

However, the idea to automate introducing changes into `config.yaml` sounds quite alarming for me. If feels like we are making the user interface unstable (because `config.yaml` is actually the major UI of the model)  which is a famous architectural anti-pattern and must be avoided.

According to the user experience we have observed, the only use case when such automation may be done in a safe way, is when `config.yam` doesn't exist. Exactly that is a case addressed by the initial configuration. So, I agree with @davide-f that the changes would be better placed under `if` condition above. What do you think?

Reply:
hey @ekatef! Thanks for the review :) Perhaps I did not explain myself too well. So, there will be now breaking changes to the workflow. As well, the workflow will not automatically change the config.yaml. It is just the order and config updates that changes. Instead of one config files, the workflow now uses multiples to ensure a full set of defaults and allow the user to make use of config diffs.  The user can now use 

* the config.yaml  as a pure "diff" config to the defaults which only contains effective changes set by the user
* the `snakemake --configfile <myconfig.yaml>`  options

See for example the new [config.tutorial](https://github.com/open-energy-transition/pypsa-earth/blob/refactor-test-and-ci/config.tutorial.yaml). It only contains effective diffs to the default. This makes it much easier to maintain both the default and tutorial config and allows to actually see what are the effective changes to defaults coming from the config.tutorial.   

Old `config.yaml` file will still work. In case there is no `config.yaml` I added a function `copy_default_files()` which sets an empty config.yaml in place as a placeholder (see some lines below). The only exception is the settings of the solver options which anyway comes with the linopy integration which are necesary to allow for config updates.

If you want we can also chat about it later, if that suits :)